### ISSUE TITLE
[Operator] Why Cumsum axis is not an attribute?

### ISSUE THREAD

**Issue by masahi** [onnx/onnx#3420]
Hi, I'm wondering why [Cumsum op](https://github.com/onnx/onnx/blob/master/docs/Operators.md#CumSum), introduced in https://github.com/onnx/onnx/pull/2030, requires `axis` to be passed as input, rather than being an attribute? Other ops that take an axis, such as Reduce*, Scatter, Gather, Split, Concat etc all seem to take axis as an attribute. Also in practice, I cannot imagine a use case where we want to have a variable axis for cumsum. 

Unless there is a good reason to have a variable axis for Cumsum, I believe this is a flaw in the spec that needs to be corrected. 

cc @jcwchen @postrational @wschin

**Comment by fdwr** [comment]
Older operators tend to take such parameters as constant attributes, but newer operators have been moving toward dynamic parameters, such as `TopK` moving `k` from constant to dynamic. I don't have the background for why cumulative summation's `axis` was set to a dynamic parameter, but I am curious (and I also wonder if this general trend of moving constant attributes to dynamic inputs will eventually make the distinction between "attributes" vs "inputs" rather obsolete, such that we *just* have inputs, the only distinction being whether they came from a `Constant` or from some other graph edge).

**Comment by masahi** [comment]
`k` parameter of `TopK` being dynamic does make sense, although it is constant in majority of cases I believe, it is not unreasonable to want `k` param being a computed expression. 

But I'd say `axis`  parameter is very different, this is not something you pick arbitrary value for at runtime. Given that other ops that needs `axis` param take it as an attribute, I don't see why `Cumsum` should be any different.

**Comment by masahi** [comment]
I see arguments vs attributes distinction doesn't really matter for typical DL frameworks or ONNXRuntime where everything is interpreted at runtime. But for DL compilers like [TVM](https://github.com/apache/tvm), whether some input is static vs dynamic has a huge difference.

**Comment by fdwr** [comment]
>whether some input is static vs dynamic has a huge difference

Indeed, the same concern applies within ONNX Runtime to the DirectML execution provider (which has a constant `DML_OPERATOR_CUMULATIVE_SUMMATION::Axis` field), but thanks to constant folding and the ability to query whether a tensor input is constant or not, the distinction between parameter types from a constant "attribute" vs a const graph initializer vs a `Const` operator all becomes moot, and we can create the GPU shader ahead of time either way. The real deciding factor is *constness*, not *attributeness*.

| Attribute | graph initializer | `Const` operator | any other operator
|-|-|-|-|
| const | const | const | dynamic

Could TVM apply the same logic at compile time, and treat all const cases equivalently? (of course, in the potential rare case that the axis *is truly* dynamic, then TVM would yield an error, but oh well)

>Given that other ops that need axis param take it as an attribute, I don't see why Cumsum should be any different.

I agree that the inconsistency is surprising, but I don't know the reason. Perhaps the opener @jeffsaremi does.

**Comment by masahi** [comment]
> Could TVM apply the same logic at compile time, and treat all const cases equivalently?

Yes, we can do the same, if `axis` parameter is provided as a constant. We can remove the argument vs attribute distinction by our ONNX frontend. But our `cumsum` op by itself expects `axis` as a compile time constant, so we would have a trouble if `axis` is truly a dynamic variable. I think we have the same concern here.

**Comment by gramalingam** [comment]
There are several questions here.
(a) Is there a use-case where axis needs to be an input instead of attribute? I don't know. But the original PR mentions the motivation was translation from TF, and it looks like axis is a "tensor" in  TF. So, perhaps the translation from tensorflow requires this?
(b) Can we change the input to attribute for CumSum? This is somewhat tricky for a spec, without a good answer to (a) from the community. In general, there have been cases where an attribute was promoted to an input, and that seems reasonable since converting a model that uses an attribute into one where it is an input is easy. But going the other direction doesn't have this justification.
(c) As for the pros/cons of attributes vs. input, the above discussion pretty much summarizes it. Hence, there has been a pull in both directions. The current recommendation is to use attributes, and promote to input only if there is a reasonable justification.

**Comment by masahi** [comment]
I don't now the convention in TF, but other op like `gather` https://www.tensorflow.org/api_docs/python/tf/gather also says `axis` is Tensor. I don't think there is something specific for cumsum.

**Comment by masahi** [comment]
I also want to given an implementer's perspective on why `axis` in cumsum needs to be static. To implement cumsum efficiently on GPU, we want to scan on the innermost axis. So we need to do transpose before/after cumsum. But if axis is dynamic, transpose also needs to take dynamic list of integers, since we don't know where the axis is. ONNX `Transpose` https://github.com/onnx/onnx/blob/master/docs/Operators.md#Transpose says the input permutation is attribute.

**Comment by masahi** [comment]
Interestingly, `Squeeze` op https://github.com/onnx/onnx/blob/master/docs/Operators.md#Squeeze takes `axes` as argument, while in the opset 11 version https://github.com/onnx/onnx/blob/master/docs/Changelog.md#Squeeze-11 `axes` was an attribute.

The change was introduced in https://github.com/onnx/onnx/pull/2955. I wonder why the change was necessary.

**Comment by masahi** [comment]
In the same PR #2955, `ReduceSum` https://github.com/onnx/onnx/blob/master/docs/Operators.md#ReduceSum was also updated to take `axes` as argument, while other `Reduce*` ops still take `axes` as an attribute till today. This inconsistency doesn't make any sense, we should fix this.

**Comment by fdwr** [comment]
@masahi: Indeed, "newer operators have been moving toward dynamic parameters". @daquexian also noted the inconsistency in #2955 "Should we also update other ops (ReduceMax/Min/...) in the ReduceX family (they all have an 'axes' attribute) so that they can be consistent with ReduceSum?".

**Comment by masahi** [comment]
Yes I generally agree with moving toward dynamic parameters, provided there is a good reason to do so for each param. In that sense, I think dynamic topk param makes sense, while I don't see why anyone would want to have dynamic `axes`.

**Comment by ashbhandare** [comment]
> Yes I generally agree with moving toward dynamic parameters, provided there is a good reason to do so for each param. In that sense, I think dynamic topk param makes sense, while I don't see why anyone would want to have dynamic `axes`.
> 
> So I'm interested to hear the motivation of #2955, in particular why the need to "Support dynamic shapes in ORT" (good) leads to the decision to add dynamic axes support in certain but not all ops (confusing).

Regarding why dynamic axes are required for certain ops like ReduceSum for supporting dynamic shapes in ORT Training:
For Training, ORT builds a gradient graph for the corresponding inference graph. If the forward graph contains a broadcast in ops like Add, Sub, Mul, etc, the gradient needs a ReduceSum on the broadcasted axes. With dynamic shapes, onnx shape inference might not always be able to populate the shapes for the entire graph. That would mean that shapes would not be available at the time of building the gradient graph, and we would only be able to determine the axes to be reduced at runtime, needing dynamic axes for ReduceSum. Essentially these changes are required for shape-independent gradient graph building, and it is a indirect consequence of using dynamic input shapes. The reasons can be summarized here:
![image](https://user-images.githubusercontent.com/14295305/118853862-122e8100-b889-11eb-9e3d-e01d5f796aa9.png)




<meta name="ProgId" content="OneNote.File">
<meta name="Generator" content="Microsoft OneNote 15">





<div style="direction:ltr;border-width:100%">

<div style="direction:ltr;margin-top:0in;margin-left:0in;width:6.825in">

<div style="direction:ltr;margin-top:0in;margin-left:0in;width:6.825in">

<div style="direction:ltr">


Op | Change | Reason
-- | -- | --
ReduceSum | 'axes' as input   instead of attribute | Gradient builder:   dymanic axes for ops with broadcast in fwd pass
Split | 'split' as input   instead of attribute | Gradient builder:   dymanic output sizes for gradient of Concat
Unsqueeze | 'axes' as input   instead of attribute | Gradient builder:   Gradient for modified&nbsp; ReduceSum   requires 'Unsqueeze' over axes that were reduced.
Squeeeze | 'axes' as input   instead of attribute | Gradient builder:   Gradient for 'Unsqueeze'



</div>

</div>

</div>

</div>









<meta name="ProgId" content="OneNote.File">
<meta name="Generator" content="Microsoft OneNote 15">





<div style="direction:ltr">


Op | Change | Reason
-- | -- | --
ReduceSum | 'axes' as input   instead of attribute | Gradient builder:   dymanic axes for ops with broadcast in fwd pass
Split | 'split' as input   instead of attribute | Gradient builder:   dymanic output sizes for gradient of Concat
Concat | Additional output:   sizes of each input along concatenated axis | To pass to Split's   'split' as input in gradient graph
Unsqueeze | 'axes' as input   instead of attribute | Gradient builder:   Gradient for modified&nbsp; ReduceSum   requires 'Unsqueeze' over axes that were reduced.
Squeeeze | 'axes' as input   instead of attribute | Gradient builder:   Gradient for 'Unsqueeze'



</div>









<meta name="ProgId" content="OneNote.File">
<meta name="Generator" content="Microsoft OneNote 15">





<div style="direction:ltr">


Op | Change | Reason
-- | -- | --
ReduceSum | 'axes' as input   instead of attribute | Gradient builder:   dymanic axes for ops with broadcast in fwd pass
Split | 'split' as input   instead of attribute | Gradient builder:   dymanic output sizes for gradient of Concat
Concat | Additional output:   sizes of each input along concatenated axis | To pass to Split's   'split' as input in gradient graph
Unsqueeze | 'axes' as input   instead of attribute | Gradient builder:   Gradient for modified&nbsp; ReduceSum   requires 'Unsqueeze' over axes that were reduced.
Squeeeze | 'axes' as input   instead of attribute | Gradient builder:   Gradient for 'Unsqueeze'



</div>

**Comment by masahi** [comment]
I want to understand why dynamic shapes make it impossible to determine the axes statically. Do you have a concrete example that doesn't work if only static axes are allowed? Also can you point me where in the ORT code base it calculates axes at runtime, based on input shapes?
